pkg/ruler/manager.go

```diff
@@ -167,6 +233,59 @@ func (r *DefaultMultiTenantManager) Start() {
 	r.rulerIsRunning.Store(true)
 }
 
+// InstanceAdded handles the addition of an instance to the alertmanager ring.
+func (r *DefaultMultiTenantManager) InstanceAdded(instance servicediscovery.Instance) {
+	if instance.InUse {
+		r.discoveryMtx.Lock()
+		defer r.discoveryMtx.Unlock()
+		level.Info(r.logger).Log("msg", "adding alertmanager instance", "addr", instance.Address)
+		r.discoveryConfigs[r.alertManagerHTTPAddress(instance)] = alertmanagerdiscovery.NewDiscoveryConfig(instance.Address)
+		r.updateNotifierConfig()
+	}
+}
+
+// InstanceRemoved handles the removal of an instance from the alertmanager ring.
+func (r *DefaultMultiTenantManager) InstanceRemoved(instance servicediscovery.Instance) {
+	r.discoveryMtx.Lock()
+	defer r.discoveryMtx.Unlock()
+	level.Info(r.logger).Log("msg", "removing alertmanager instance", "addr", instance.Address)
+	delete(r.discoveryConfigs, r.alertManagerHTTPAddress(instance))
+	r.updateNotifierConfig()
+}
+
+// InstanceChanged handles a change to an instance in the alertmanager ring.
+func (r *DefaultMultiTenantManager) InstanceChanged(instance servicediscovery.Instance) {
+	if instance.InUse {
+		r.InstanceAdded(instance)
+	} else {
+		r.InstanceRemoved(instance)
+	}
+}
+
+// alertManagerHTTPAddress returns an alert manager HTTP address for the instance.
+func (r *DefaultMultiTenantManager) alertManagerHTTPAddress(instance servicediscovery.Instance) string {
+	return "http://" + instance.Address + r.alertmanagerHTTPPrefix
+}
+
+// updateNotifierConfig builds a new notifier config and applies it to existing notifiers.
+// Must lock discoveryMtx prior to calling.
+func (r *DefaultMultiTenantManager) updateNotifierConfig() {
+	ncfg, err := buildNotifierConfig(&r.cfg, r.discoveryConfigs)
+	if err != nil {
+		level.Error(r.logger).Log("msg", "unable to build updated notifier config", "err", err)
+		return
+	}
+	r.notifierCfg = ncfg
```

Comment:
Move after taking the lock below. The `r.notifierCfg` field comment says it's guarded by `r.notifiersMtx`.

Reply:
Good call - will move that line. Re: the comment, unfortunately a comment is all I can do since a lock on `discoveryConfigsMtx` is needed, but since locks aren't re-entrant and in this case, the lock is already held prior to calling the method.